packages/gatsby/src/redux/reducers/queries.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  ActionsUnion,
+  IComponentState,
+  IGatsbyState,
+  IQueryState,
+} from "../types"
+
+type QueryId = string // page query path or static query id
+type ComponentPath = string
+type NodeId = string
+type ConnectionName = string
+
+export const FLAG_DIRTY_PAGE = 0b0001
+export const FLAG_DIRTY_TEXT = 0b0010
+export const FLAG_DIRTY_DATA = 0b0100
+
+export const FLAG_ERROR_BABEL = 0b0001
+
+const initialState = (): IGatsbyState["queries"] => {
+  return {
+    byNode: new Map<NodeId, Set<QueryId>>(),
+    byConnection: new Map<ConnectionName, Set<QueryId>>(),
+    trackedQueries: new Map<QueryId, IQueryState>(),
+    trackedComponents: new Map<ComponentPath, IComponentState>(),
+  }
+}
+
+const initialQueryState = (): IQueryState => {
+  return {
+    dirty: -1, // unknown, must be set right after init
+  }
+}
+
+const initialComponentState = (): IComponentState => {
+  return {
+    componentPath: ``,
+    query: ``,
+    pages: new Set<QueryId>(),
+    errors: 0,
+    // TODO: staticQueries: new Set<QueryId>()
+  }
+}
+
+/**
+ * Tracks query dirtiness. Dirty queries are queries that:
+ *
+ * - depend on nodes or node collections (via `actions.createPageDependency`) that have changed.
+ * - have been recently extracted (or their query text has changed)
+ * - belong to newly created pages (or pages with modified context)
+ *
+ * Dirty queries must be re-ran.
+ */
+export function queriesReducer(
+  state: IGatsbyState["queries"] = initialState(),
+  action: ActionsUnion
+): IGatsbyState["queries"] {
+  switch (action.type) {
+    case `DELETE_CACHE`:
+      return initialState()
+
+    case `CREATE_PAGE`: {
+      const { path, componentPath } = action.payload
+      if (!state.trackedQueries.has(path) || action.contextModified) {
+        const query = registerQuery(state, path)
+        query.dirty = setFlag(query.dirty, FLAG_DIRTY_PAGE)
+      }
+      registerComponent(state, componentPath).pages.add(path)
+      return state
+    }
+    case `DELETE_PAGE`: {
+      const { path, componentPath } = action.payload
+      const component = state.trackedComponents.get(componentPath)
+      if (component) {
+        component.pages.delete(path)
+      }
+      state.trackedQueries.delete(path)
+      return state
+    }
+    case `QUERY_EXTRACTED`: {
+      // Note: this action is called even in case of
+      // extraction error or missing query (with query === ``)
+      // TODO: use hash instead of a query text
+      const { componentPath, query } = action.payload
+      const component = registerComponent(state, componentPath)
+      if (component.query !== query) {
+        // Invalidate all pages associated with a component when query text changes
+        component.pages.forEach(queryId => {
+          const query = state.trackedQueries.get(queryId)
+          if (query) {
+            query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+          }
+        })
+        component.query = query
+      }
```

Comment:
Does this change behaviour from current?

As in - if user have babel error and then fix it without changing query itself - this would rerun query (unnecessarily?). Do we do this in master too?

Reply:
You are right (as always). Addressed here: https://github.com/gatsbyjs/gatsby/pull/27504/commits/089f898be721787e852f07385b7c89f2f5fb151b

Also tested locally and now it won't re-run queries for recovered components (unless they've actually changed).